onnx/defs/controlflow/defs.cc

```diff
@@ -774,4 +774,102 @@ ONNX_OPERATOR_SET_SCHEMA(
+    some parameters:
+     
+     - The initial learning-rate "R".
+     - The update count "T". That is, the number of training iterations conducted.
+     - A Frobenius norm regularization coefficient "Lambda".
+     - A learning-rate decay factor per iteration "D".
+     - A small constant "Eps" to avoid dividing-by-zero. 
+
+    At each ADAGRAD iteration, the optimized tensor variables are moved along a direction
+    computed based on their estimated gradient and accumulated squared gradient. Assume
+    that only a single tensor "X" is updated by this operator. We need the value of "X",
+    its gradient "G", and its accumulated squared gradient "H". Consequently, if "X" is
+    the only one tensor to be optimized, variables in this operator's input list are
+    sequentially "R", "T", "D", "Eps", "X", "G", and "H". Also, the corresponding output
+    tensors are the new value of "X" (called "X_new"), and the new accumulated squared 
+    gradient (called "H_new"). Those outputs are computed from the given inputs following
+    the pseudo code below.
+
+    Let "+", "-", "*", and "/" are all element-wise operations with numpy-style broadcasting.
+    The pseudo code to compute those outputs is:
+
+      // Compute a scalar learning-rate factor. If X is never updated, T is 0.
+      r = R / (1 + T * D);
+
+      // Add gradient of 0.5 * Lambda * ||X||_F^2, where ||X||_F is the Frobenius norm.
+      G_regularized = Lambda * X + G;
+
+      // Compute new accumulated squared gradient.
+      H_new = H + G_regularized * G_regularized;
+
+      // Compute the adaptive part of per-coordinate learning rate. Note that Sqrt(...)
+      // compute square root element-wisely.
+      H_adaptive = Sqrt(H_new) + Eps
+
+      // Compute the new value of "X".
+      X_new = X - r * G_regularized / H_adaptive;
+
+    If one assign this operators to optimize multiple inputs, for example, "X_1" and "X_2". The same
+    pseudo code would be extended to handle all tensors jointly. More specifically, we can view "X" as a
+    concatenation of "X_1" and "X_2" (of course, their gradient and accumulate gradient should
+    be concatenated too) and then just execute our pseudo code.
+
+    Note that ADAGRAD was first proposed in http://jmlr.org/papers/volume12/duchi11a/duchi11a.pdf.
+    In that reference paper, this operator is a spacial case of the Figure 1's composite mirror
+    descent update.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Adagrad,
+    10,
+    OpSchema()
+        .SetDoc(Adagrad_ver10_doc)
+        .Input(0, "R", "The initial learning rate.", "T1")
+        .Input(1, "D", "The decay factor of learning rate after one update.", "T1")
+        .Input(2, "T", "The update count of \"X\". It should be a scalar.", "T3")
+        .Input(3, "Lambda", "Regularization coefficient of 0.5 * Lambda * ||X||_F^2.", "T1")
+        .Input(4, "Eps", "Small scalar to avoid dividing by zero.", "T1")
+        .Input(
+            5,
+            "inputs",
```

Comment:
Gradient and accumulated squared gradient should be optional because a model can come without backward graph.

Reply:
Maybe I am wrong. It causes ambiguation between [X, G] and [X_1, X_2].